### PR TITLE
Set ActiveStatus when retention size config option is valid

### LINKED ISSUE #560: Stuck on BlockedStatus even after `maximum_retention_size` config option changed to valid
If we provide an invalid string to `maximum_retention_size`, the charm goes into BlockedStatus. Then, when it is corrected, there is no code in place to revert the status back to active.

https://github.com/canonical/prometheus-k8s-operator/blob/c03298b7736f2aa3623446c86014263a94bdd8c4/src/charm.py#L678-L683

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
BlockedStatus from an invalid retention size config option was never cleared.

## Solution
- Add a TypedDict for per-component status setting.
- Use the per-component status dict to set status centrally in `_configure`.

Fixes #560.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -563,13 +575,9 @@ def _configure(self, _):
...
-        if (
-            isinstance(self.unit.status, BlockedStatus)
-            and self.unit.status not in early_return_statuses.values()
-        ):
-            return
-
-        self.unit.status = ActiveStatus()
+        self.unit.status = StatusBase._get_highest_priority(
+            cast(List[StatusBase], list(self.status.values()))
+        )
```
Comment (sed-i) [inline src/charm.py:580]: @benhoyt, just drawing your attention to this status-related (ab)use, if you have a moment to take a look.

Currently there's a mix between centralized and decentralized status setting, and collect-status can't be used without major refactoring.

An alternative could be using the collect-status machinery to commit all values inside `self.status`.

**Review by PietroPasotti — APPROVED** [review]
lgtm

File: src/charm.py (hunk 1 above)
Comment (benhoyt) [reply src/charm.py:580]: Yeah, that sounds like a good direction to me too.

Additionally, I'd prefer charms didn't use private (underscore-prefixed) Ops methods like `StatusBase._get_highest_priority`. That said, I think that one's been useful enough for people doing various things with status that maybe we should make it public, either as `StatusBase.get_highest_priority` or another name. If you're interested in that, please open an issue for us to consider.

File: src/charm.py [hunk 2]
```diff
@@ -91,6 +93,26 @@ class ConfigError(Exception):
...
+    """Per-component status holder."""
+
+    # These are going to go into stored state, so we must use marshallable objects.
+    # They are passed to StatusBase.from_name().
+    retention_size: Tuple[str, str]
+    k8s_patch: Tuple[str, str]
+    config: Tuple[str, str]
+
+
+def to_tuple(status: StatusBase) -> Tuple[str, str]:
```
[the remark this answers is not in the document]
Comment (sed-i) [reply src/charm.py:106]: Could really appreciate rust traits here :)

I agree in general but here I think a shorter name is a better tradeoff: used only in this file, intention is immediately clear from either the definition or the usage, static checker would complain if misused.

**Pull request merged**